### PR TITLE
Improve error checking in mesh generators

### PR DESCRIPTION
**Pull request by aprilnovak**
This adds extra error checks on user input for boundary IDs/names in the mesh generators. This does not address every occurence where the user provides a boundary/sideset ID/name, but I thought it better to split into smaller PRs since there's a decent amount here already.

Refs #22117

### CODE REVIEW

**Comment by GiudGiud** [comment]
Seems there's a failure in BISON. Do you have access and do you want to look into it or should we ping that team?

File: framework/src/utils/MooseMeshUtils.C [hunk 1]
```diff
@@ -320,4 +320,34 @@ getNextFreeSubdomainID(MeshBase & input_mesh)
...
+{
+  std::set<SubdomainID> mesh_blocks;
+  input_mesh.subdomain_ids(mesh_blocks);
+  return mesh_blocks.count(id) && (id != Moose::INVALID_BLOCK_ID);
+}
+
+bool
+hasSubdomainName(MeshBase & input_mesh, const SubdomainName & name)
+{
+  auto id = getSubdomainID(name, input_mesh);
```
Comment (GiudGiud) [inline framework/src/utils/MooseMeshUtils.C:335]:
```suggestion
  const auto id = getSubdomainID(name, input_mesh);
```

File: framework/src/utils/MooseMeshUtils.C [hunk 2]
```diff
@@ -320,4 +320,34 @@ getNextFreeSubdomainID(MeshBase & input_mesh)
...
+{
+  const libMesh::BoundaryInfo & boundary_info = input_mesh.get_boundary_info();
+  std::set<libMesh::boundary_id_type> boundary_ids = boundary_info.get_boundary_ids();
+  return boundary_ids.count(id) && (id != Moose::INVALID_BOUNDARY_ID);
+}
+
+bool
+hasBoundaryName(MeshBase & input_mesh, const BoundaryName & name)
+{
+  auto id = getBoundaryID(name, input_mesh);
```
Comment (GiudGiud) [inline framework/src/utils/MooseMeshUtils.C:350]:
```suggestion
  const auto id = getBoundaryID(name, input_mesh);
```

File: test/tests/meshgenerators/boundary_deletion_generator/tests [hunk 3]
```diff
@@ -1,11 +1,20 @@
...
     design = 'meshgenerators/BoundaryDeletionGenerator.md'
     issues = '#11174'
-  [../]
+  []
+  [invalid_boundary]
+    type = RunException
+    input = boundary_deletion.i
+    cli_args = "Mesh/boundary_removal/boundary_names='missing'"
+    expect_err = "The boundary 'missing' was not found in the mesh"
+    requirement = "The system shall error if deleting a non-existent boundary"
```
Comment (GiudGiud) [inline test/tests/meshgenerators/boundary_deletion_generator/tests:16]:
```suggestion
    requirement = "The system shall error if attempting to delete a non-existent boundary"
```

File: test/tests/meshgenerators/parsed_generate_sideset/tests [hunk 4]
```diff
@@ -24,4 +24,22 @@
     design = 'meshgenerators/ParsedGenerateSideset.md'
     issues = '#15651'
   [../]
+
+  [invalid_included_subdomains]
+    type = RunException
+    input = parsed_generate_sideset.i
+    cli_args = 'Mesh/sideset/included_subdomains="missing"'
+    expect_err = "The block 'missing' was not found in the mesh"
+    requirement = 'The system shall error if the included subdomains do not exist in the mesh.'
```
Comment (GiudGiud) [inline test/tests/meshgenerators/parsed_generate_sideset/tests:33]:
```suggestion
    requirement = 'The system shall error if one of the included subdomains does not exist in the mesh.'
```

File: test/tests/meshgenerators/parsed_generate_sideset/tests [hunk 5]
```diff
@@ -24,4 +24,22 @@
...
+    issues = '#22117'
+  []
+
+  [invalid_included_neighbors]
+    type = RunException
+    input = parsed_generate_sideset.i
+    cli_args = 'Mesh/sideset/included_neighbors="missing"'
+    expect_err = "The block 'missing' was not found in the mesh"
+    requirement = 'The system shall error if the included neighbors do not exist in the mesh.'
```
Comment (GiudGiud) [inline test/tests/meshgenerators/parsed_generate_sideset/tests:42]:
```suggestion
    requirement = 'The system shall error if one of the included neighbors does not exist in the mesh.'
```

File: test/tests/meshgenerators/parsed_subdomain_mesh_generator/tests [hunk 6]
```diff
@@ -11,4 +11,12 @@
     design = 'meshgenerators/ParsedSubdomainMeshGenerator.md'
     issues = '#11640'
   [../]
+  [invalid_excluded_subdomain]
+    type = RunException
+    input = parsed_subdomain_mg.i
+    cli_args = "Mesh/subdomains2/excluded_subdomains='missing'"
+    expect_err = "The block 'missing' was not found in the mesh"
+    requirement = 'The system shall error if subdomains are specified that do not exist in the mesh.'
```
Comment (GiudGiud) [inline test/tests/meshgenerators/parsed_subdomain_mesh_generator/tests:19]:
```suggestion
    requirement = 'The system shall error if a specified subdomain does not exist in the mesh.'
```

File: test/tests/meshgenerators/refine_block_generator/tests [hunk 7]
```diff
@@ -30,4 +30,12 @@
       detail = 'multiple blocks within a multi-domain mesh, with neighbor refinement enabled'
     []
   []
+
+  [invalid_block]
+    type = RunException
+    input = test_single.i
+    cli_args = 'Mesh/refine/block="missing"'
+    expect_err = "The block 'missing' was not found within the mesh"
+    requirement = "The system shall error if refining a block that does not exist in the mesh."
```
Comment (GiudGiud) [inline test/tests/meshgenerators/refine_block_generator/tests:39]:
```suggestion
    requirement = "The system shall error if attempting to refine a block that does not exist in the mesh."
```

File: test/tests/meshgenerators/subdomain_bounding_box_generator/tests [hunk 8]
```diff
@@ -54,4 +54,16 @@
     mesh_mode = 'REPLICATED'
     recover = false
   [../]
+
+  [invalid_restricted_subdomain]
+    type = RunException
+    input = subdomain_bounding_box_generator_restricted.i
+    cli_args = "Mesh/subdomains2/restricted_subdomains='missing'"
+    expect_err = "The block 'missing' was not found in the mesh"
+    requirement = 'The system shall error if a restricted subdomain does not exist in the mesh'
```
Comment (GiudGiud) [inline test/tests/meshgenerators/subdomain_bounding_box_generator/tests:63]:
```suggestion
    requirement = 'The system shall error if a restricted subdomain of the bounding box subdomain generation does not exist in the mesh'
```

**Review by GiudGiud — COMMENTED** [review]
some nitpicking

**Comment by aprilnovak** [comment]
I have access to BISON, I can address the error

**Comment by GiudGiud** [comment]
No I dont think they are related, wont even build.
This one maybe?
```
dgkernels/dg_block_restrict.2D_test: /data/civet0/build/moose/test/tests/dgkernels/dg_block_restrict/2d_dg_diffusion_block_restrict.i:27: (Mesh/boundaries/boundaries):
dgkernels/dg_block_restrict.2D_test:     The boundary 'bottom' was not found in the mesh
```
in distributed threads https://civet.inl.gov/job/1183434/

**Comment by aprilnovak** [comment]
Yes, I think it's related. I didn't account for the possibility that a sideset/subdomain might only exist on a subset of the ranks. I think I've corrected this now.

**Review by GiudGiud — APPROVED** [review]
looks great, much better for user experience!

**Comment by aprilnovak** [comment]
Fairly sure the BISON error is unrelated. I can't see the ARM Mac failures for some reason, it just pinwheels while loading and downloading the .tar.gz gives an empty file.

**Comment by GiudGiud** [comment]
Indeed unrelated. ARM passes now, got fixed

**Comment by aprilnovak** [comment]
Thanks for the review!

**Pull request merged**